src/apply.jl

```diff
@@ -25,16 +25,43 @@ function ITensors.apply(
     if ortho
       ψ = orthogonalize(ψ, v⃗[1])
     end
-    oψᵥ = apply(o, ψ[v⃗[1]] * ψ[v⃗[2]])
-    ψᵥ₁, ψᵥ₂ = factorize(
-      oψᵥ, inds(ψ[v⃗[1]]); cutoff, maxdim, tags=ITensorNetworks.edge_tag(e)
-    )
+
+    #Check whether to do a memory efficient QR first (do both sites combined for now, although we could split if we wish)
+    #if no maxdim is provided just do it based on inds counting, if it is provided do it based on whichever takes less memory
+    d1, d2 = prod(dim.(inds(ψ[v⃗[1]]; tags="Site"))),
+    prod(dim.(inds(ψ[v⃗[2]]; tags="Site")))
```

Comment:
We should always avoid using tags for this kind of thing (at this point I only use tags for printing, defining operators, and for infinite tensor networks).

Probably you can use some index set logic and graph analysis.

Reply:
Okay, I have written it now so that it does it based on the indices unique to `ψ[v⃗[1]]` compared to its neighbours.

I.e:

```
v1_siteinds = setdiff(inds(ψ[v⃗[1]]), flatten(unique(inds.([ψ[v] for v in neighbors(ψ, v⃗[1])]))))
v2_siteinds = setdiff(inds(ψ[v⃗[2]]), flatten(unique(inds.([ψ[v] for v in neighbors(ψ, v⃗[2])]))))
d1, d2 = prod(dim.(v1_siteinds)), prod(dim.(v2_siteinds))
```

Was that along the lines you were thinking?

Reply:
I thought the overall logic could be a bit simpler, like:
```julia
dim_v1 = dim(uniqueinds(ψ[v⃗[1]], ψ[v⃗[2]])
dim_v2 = dim(uniqueinds(ψ[v⃗[2]], ψ[v⃗[1]])
dim_shared = dim(commoninds(ψ[v⃗[1]], ψ[v⃗[2]]))
```
and then you can choose whether or not to QR based on those three.

So it doesn't seem like you should need the site/external indices in particular. However if you did (or need it for something else), there is the function `uniqueinds(tn::AbstractITensorNetwork, vertex)`: https://github.com/mtfishman/ITensorNetworks.jl/blob/7bdcc89b672510cdcab6e4bb7618f58c074db586/src/abstractitensornetwork.jl#L230-L232

Reply:
Agreed this should make it simpler. But isn't the condition for whether to do a QR: `dim_v1*dim_v2 < d*d*dim_shared*dim_shared`? So there is the need to work out `d`, the dimension of the physical index?

Reply:
Ah right. So in this case the "site indices" are really defined as the ones that are shared with the gate, so we should use the gate in the logic, i.e.:
```julia
outer_dim_v1 = dim(uniqueinds(ψ[v⃗[1]], o, ψ[v⃗[2]]))
dim_v1 = dim(commoninds(ψ[v⃗[1]], o))
shared_dim = dim(commoninds(ψ[v⃗[1]], ψ[v⃗[2]]))
```
and the comparison is `outer_dim_v1 > shared_dim * dim_v1`. Maybe I got the details wrong but hopefully you get the point that we should use the graph connectivity and set index logic as much as possible.

Reply:
Maybe I have it wrong about using `o` to determine the site indices vs. `uniqueinds(ψ, v⃗[1])`, those can be different if the gate/operator is only applied to a subset of the site indices of a vertex, like applying a gate to a density matrix.

Reply:
I think what you have there makes the most sense? If the gate is only being applied to a subset of the site indices of a vertex then we should treat the remaining site indices as ones we want to put onto Q and not R in the decomposition? Hence the lines you wrote there would work?